### PR TITLE
[Airflow-1413] Fix FTPSensor failing on error message with unexpected text.

### PR DESCRIPTION
**Pull request by mdziemianko**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below fixing the issue outlined in https://issues.apache.org/jira/browse/AIRFLOW-1413

The FTPSensor relies on an error message returned by the FTPHook. The expected message is "Can't check for file existence" and is used to check for errors in following way:
try:
    hook.get_mod_time(self.path)
except ftplib.error_perm as e:
    error = str(e).split(None, 1)
    if error[1] != "Can't check for file existence":
        raise e
However on my system (Linux/Antregos) this message text is different, leading to inevitable dag termination. Moreover the actual format of the exception on my system is different and the split returns '-' instead of the actual message.

This patch changes the operation to inspect error code rather than message text.
It also adds option to ignore certain classes of errors such as Host Unavailable that are recoverable, thus the performed action can and should be retried according to ftp spec. The default setting is to fail on all errors in order to preserve current behaviour.

Additional test cases added in contrib/sensors/test_ftp_sensor.py

### CODE REVIEW

**Comment by ron819** [comment]
@Fokko

File: airflow/contrib/sensors/ftp_sensor.py [hunk 1]
```diff
@@ -26,33 +28,64 @@
...
         return FTPHook(ftp_conn_id=self.ftp_conn_id)
 
+    def _get_error_code(self, e):
+        """Extract error code from ftp exception"""
+        try:
+            matches = self.error_code_pattern.match(str(e))
+            code = int(matches.group(0))
+            return code
+        except ValueError:
+            return None
```
Comment (ashb) [inline airflow/contrib/sensors/ftp_sensor.py]: Returning None here feels like a way to silently mask errors in exceptional cases - returning e again here feels safer.

File: airflow/contrib/sensors/ftp_sensor.py [hunk 2]
```diff
@@ -26,33 +28,64 @@
...
-    def __init__(self, path, ftp_conn_id='ftp_default', *args, **kwargs):
+    def __init__(
+            self,
+            path,
+            ftp_conn_id='ftp_default',
+            fail_on_transient_errors=True,
+            *args,
+            **kwargs):
+        """
+        Create a new FTP sensor
```
Comment (ashb) [inline airflow/contrib/sensors/ftp_sensor.py:48]:
```suggestion
        Create a new FTP sensor
        
```

There needs to be a blank line between the description and the params for this to render correctly

File: airflow/contrib/sensors/ftp_sensor.py [hunk 3]
```diff
@@ -26,33 +28,64 @@
...
+
     def poke(self, context):
         with self._create_hook() as hook:
             self.log.info('Poking for %s', self.path)
             try:
                 hook.get_mod_time(self.path)
             except ftplib.error_perm as e:
-                error = str(e).split(None, 1)
-                if error[1] != "Can't check for file existence":
+                logging.info('Ftp error encountered: %s', str(e))
```
Comment (ashb) [inline airflow/contrib/sensors/ftp_sensor.py]:
```suggestion
                self.log.info('Ftp error encountered: %s', str(e))
```

(And remove the import of `logging` at the top)

**Review by ashb — CHANGES_REQUESTED** [review]

**Comment by ashb** [comment]
(Better late than never with a review right? Sorry it took sooooo long!)

**Comment by mdziemianko** [comment]
> (Better late than never with a review right? Sorry it took sooooo long!)

lol:) I forgot about this PR... will do the changes. Thanks.

**Review by ashb — APPROVED** [review]

File: tests/contrib/sensors/test_ftp_sensor.py [hunk 4]
```diff
@@ -66,6 +71,27 @@ def test_poke_fails_due_error(self):
...
+        op = FTPSensor(path="foobar.json", ftp_conn_id="bob_ftp",
+                       task_id="test_task", fail_on_transient_errors=False)
+
+        self.hook_mock.get_mod_time.side_effect = \
+            [error_perm("434: Host unavailable"), None]
+
+        self.assertFalse(op.poke(None))
+        self.assertTrue(op.poke(None))
 
 if __name__ == '__main__':
```
Comment (ashb) [inline tests/contrib/sensors/test_ftp_sensor.py:97]: Tests are now failing the style check here with:

> ./tests/contrib/sensors/test_ftp_sensor.py:96:1: E305 expected 2 blank lines after class or function definition, found 1

```suggestion

if __name__ == '__main__':
```

**Comment by Fokko** [comment]
Restarted the failed job to check if the error is transient

**Comment by mdziemianko** [comment]
seems like all checks passed...:)

**Review by Fokko — APPROVED** [review]
Thanks @mdziemianko

**Pull request merged**